connect/cli/plugins/project/extension_helpers.py

```diff
@@ -41,3 +51,110 @@ def bootstrap_extension_project(data_dir: str):
+        )
+
+    extension_dict = data['tool']['poetry']['plugins']['connect.eaas.ext']
+    if not isinstance(extension_dict, dict):
+        raise ClickException(
+            'The extension definition on [tool.poetry.plugins."connect.eaas.ext"] `pyproject.toml` section '
+            'is not well configured. It should be as following: "extension" = "your_package.extension:YourExtension"',
+        )
+    if 'extension' not in extension_dict.keys():
+        raise ClickException(
+            'The extension definition on [tool.poetry.plugins."connect.eaas.ext"] `pyproject.toml` section '
+            'does not have "extension" defined. Reminder: "extension" = "your_package.extension:YourExtension"',
+        )
+    return extension_dict
+
+
+def _entrypoint_validations(project_dir, extension_dict):
+    package_name = extension_dict['extension'].rsplit('.', 1)[0]
+    descriptor_file = os.path.join(f'{project_dir}/{package_name}', 'extension.json')
+    ext_class = next(pkg_resources.iter_entry_points('connect.eaas.ext', 'extension'), None)
+    if not ext_class:
+        raise ClickException('\nThe extension could not be loaded, Did you execute `poetry install`?')
+
+    CustomExtension = ext_class.load()
+    if not inspect.isclass(CustomExtension):
+        raise ClickException(f'\nThe extension class {CustomExtension} does not seem a class, please check it')
+
+    all_methods = CustomExtension.__dict__
+    methods = [method for method in all_methods.keys() if not method.startswith('__')]
+
+    try:
+        ext_descriptor = json.load(open(descriptor_file, 'r'))
+    except json.JSONDecodeError:
+        raise ClickException(
+            '\nThe extension descriptor file `extension.json` could not be loaded.',
+        )
+
+    capabilities = ext_descriptor['capabilities']
+
+    errors = _have_capabilities_proper_stats(capabilities)
+    if errors:
+        raise ClickException(f'Capability errors: {errors}')
+
+    if not _have_methods_proper_capabilities(methods, capabilities):
+        raise ClickException(
+            '\nThere is some mismatch between capabilities on `extension.json` '
+            'and the methods defined on your extension class on `extension.py`, '
+            'please check it.',
+        )
+
+    _have_methods_proper_type(CustomExtension, capabilities)
+
+
+def _have_methods_proper_type(cls, capabilities):
+    guess_async = [
+        inspect.iscoroutinefunction(getattr(cls, CAPABILITY_METHOD_MAP[name]))
+        for name in capabilities.keys()
+    ]
+    if all(guess_async):
+        return
+    if not any(guess_async):
+        return
+
+    raise ClickException('An Extension class can only have sync or async methods not a mix of both.')
+
+
+def _have_capabilities_proper_stats(capabilities):
+    errors = []
+    for capability, stats in capabilities.items():
+        if capability == 'product_action_execution' or capability == 'product_custom_event_processing':
+            if stats:
```

Comment:
In this case, we should ensure that the value of these two capabilities is an empty list. A null value can cause issues.

Reply:
Done!